### PR TITLE
Added support for Transactional Memory Atomic extension

### PR DESCRIPTION
**Pull request by shubhamnarlawar77**

### CODE REVIEW

**Comment by shubhamnarlawar77** [comment]
@marxin : review request

**Comment by shubhamnarlawar77** [comment]
1. Generating transactional atomic block at start of function so that there are no jumping of goto statements in the block.
2. Used naming convention similar to Csmith like cnt for count, blk for block, etc

File: src/RandomProgramGenerator.cpp [hunk 1]
```diff
@@ -831,6 +832,17 @@ main(int argc, char **argv)
 			continue;
 		}
 
+		if (strcmp (argv[i], "--trans-memory-atomic") == 0) {
+			CGOptions::trans_memory_atomic(true);
+			CGOptions::volatiles(false);
```
Comment (marxin) [inline src/RandomProgramGenerator.cpp:837]: Why do you set `volatiles`?

File: src/Function.cpp [hunk 2]
```diff
@@ -875,6 +875,16 @@ GenerateFunctions(void)
 			ERROR_RETURN();
 		}
 	}
+	if(CGOptions::trans_memory_atomic()){
+		// Total number of transaction block to be generated
+		int atomic_blk_cnt = rnd_upto(FuncListSize()) + 1 ;
```
Comment (marxin) [inline src/Function.cpp]: I don't like. I would rather set a small probability that will drive the emission of `__transaction_atomic`. You then don't need to randomly select a function, etc.

**Review by marxin — CHANGES_REQUESTED** [review]

File: src/RandomProgramGenerator.cpp (hunk 1 above)
Comment (shubhamnarlawar77) [reply src/RandomProgramGenerator.cpp:837]: Transaction block do not accept volatile variable in it.

File: src/Function.cpp (hunk 2 above)
Comment (shubhamnarlawar77) [reply src/Function.cpp]: Done. Please check it.

File: src/Function.cpp [hunk 3]
```diff
@@ -875,6 +875,14 @@ GenerateFunctions(void)
 			ERROR_RETURN();
 		}
 	}
+	if(CGOptions::trans_memory_atomic()){
+		for (int i = 0; i < FuncListSize(); i++) {
+			if(rnd_flipcoin(TransAtomicProb)){
+				if (!FuncList[i]->blocks[0]->atomic_blk)
+						FuncList[i]->blocks[0]->atomic_blk = true;
```
Comment (marxin) [inline src/Function.cpp]: Still I don't like this.

File: src/Block.cpp [hunk 4]
```diff
@@ -295,6 +295,13 @@ Block::Output(std::ostream &out, FactMgr* fm, int indent) const
 	ss << "block id: " << stm_id;
 	output_comment_line(out, ss.str());
 
+	if (atomic_blk){
```
Comment (marxin) [inline src/Block.cpp]: Can you please make `rnd_flipcoin` here?

**Review by marxin — CHANGES_REQUESTED** [review]

File: src/RandomProgramGenerator.cpp (hunk 1 above)
Comment (marxin) [reply src/RandomProgramGenerator.cpp:837]: Ok, thanks for explanation.

File: src/Function.cpp (hunk 3 above)
Comment (shubhamnarlawar77) [reply src/Function.cpp]: Should I remove all of above? I did not understand what you want.

Comment (marxin) [reply src/Function.cpp]: Yes, remove it and make the random decision in `Block::Output`.

File: src/Block.cpp (hunk 4 above)
Comment (shubhamnarlawar77) [reply src/Block.cpp]: Done. Now, instead of creating atomic block at start, they are created anywhere in a function.

Comment (marxin) [reply src/Block.cpp]: Which is much nicer!

File: src/Probabilities.cpp [hunk 5]
```diff
@@ -510,6 +510,9 @@ Probabilities::set_single_name_maps()
 
 	//for choosing type attributes
 	set_single_name("type_attr_flag", pTypeAttrProb);
+
+	//for generating transaction atomic blocks
+	set_single_name("type_attr_flag", pTransAtomicProb);
```
Comment (marxin) [inline src/Probabilities.cpp]: A typo?

File: src/Block.cpp (hunk 4 above)
Comment (shubhamnarlawar77) [reply src/Block.cpp]: This one! Is pull request ready to merge?

File: src/Probabilities.cpp [hunk 6]
```diff
@@ -554,6 +557,7 @@ Probabilities::initialize_single_probs()
 	m[pFuncAttrInline] = 50;
 	m[pFuncAttrProb] = 30;
 	m[pTypeAttrProb] = 50;
+	m[pTransAtomicProb] = 50;
```
Comment (marxin) [inline src/Probabilities.cpp]: This is very high, I would set it to 3% or so?

**Review by marxin — CHANGES_REQUESTED** [review]

**Comment by marxin** [comment]
For the next time, please select better names for branch names.

File: src/Probabilities.cpp (hunk 5 above)
Comment (shubhamnarlawar77) [reply src/Probabilities.cpp]: My bad. Corrected the typo.

File: src/Probabilities.cpp (hunk 6 above)
Comment (shubhamnarlawar77) [reply src/Probabilities.cpp]: Done.

**Comment by shubhamnarlawar77** [comment]
> For the next time, please select better names for branch names.

Sure. Base branch name will be same "csmith-gcc-v2". 
Pull request branch name will be based on the name extension!

Is pull request fine to merge now?

**Comment by marxin** [comment]
I'm reading the documentation and there is also `__transaction_cancel` directive. Have you considered supporting that?

**Comment by shubhamnarlawar77** [comment]
> I'm reading the documentation and there is also `__transaction_cancel` directive. Have you considered supporting that?

No. But I can try to add that also. Is process to add `__transaction_cancel` same as `atomic`?

**Comment by marxin** [comment]
> No. But I can try to add that also. Is process to add `__transaction_cancel` same as `atomic`?

I don't know. Please read the documentation.

**Comment by shubhamnarlawar77** [comment]
> > No. But I can try to add that also. Is process to add `__transaction_cancel` same as `atomic`?
> 
> I don't know. Please read the documentation.

It is used to roll back the transaction. Hence, if `atomic` block is set, if `rnd_flipcoin()` is passed by `cancel` then `__transaction_cancel` is printed. Please check it.

**Comment by marxin** [comment]
> It is used to roll back the transaction. Hence, if `atomic` block is set, if `rnd_flipcoin()` is passed by `cancel` then `__transaction_cancel` is printed. Please check it.

Well, the result would be that the transaction is canceled. That can break a generated csmith program which tend to be a valid code, right? Thus I would leave `__transaction_cancel` for now.

**Comment by shubhamnarlawar77** [comment]
> > It is used to roll back the transaction. Hence, if `atomic` block is set, if `rnd_flipcoin()` is passed by `cancel` then `__transaction_cancel` is printed. Please check it.
> 
> Well, the result would be that the transaction is canceled. That can break a generated csmith program which tend to be a valid code, right? Thus I would leave `__transaction_cancel` for now.

Sure. Updated the pull request!

**Comment by marxin** [comment]
It's fine, thank you.

**Comment by shubhamnarlawar77** [comment]
Should I merge it now?

**Comment by marxin** [comment]
Yes please.

**Pull request merged**